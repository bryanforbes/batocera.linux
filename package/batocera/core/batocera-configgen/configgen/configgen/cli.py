from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
