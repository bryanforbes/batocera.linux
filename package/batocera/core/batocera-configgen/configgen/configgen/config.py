from __future__ import annotations

import logging
import xml.etree.ElementTree as ET
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, ClassVar, Literal, Self, TypeVar, overload

import yaml

from .batoceraPaths import BATOCERA_SHADERS, DEFAULTS_DIR, ES_SETTINGS, USER_SHADERS
from .utils.missing import MISSING, MissingType

if TYPE_CHECKING:
    from pathlib import Path

    from .cli import Options
    from .settings.unixSettings import UnixSettings

_T = TypeVar('_T')
_TV = TypeVar('_TV')
_FV = TypeVar('_FV')

_logger = logging.getLogger(__name__)


# adapted from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
def _dict_merge(destination: dict[str, Any], source: Mapping[str, Any]) -> None:
    """Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param destination: dict onto which the merge is executed
    :param source: dict merged into destination
    :return: None
    """
    for key, value in source.items():
        if key in destination and isinstance(destination[key], dict) and isinstance(value, Mapping):
            _dict_merge(destination[key], value)
        else:
            destination[key] = value


def _load_defaults(system: str, default_file: Path, default_arch_file: Path, /) -> dict[str, Any]:
    with default_file.open() as f:
        defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader)

    with default_arch_file.open() as f:
        arch_defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader) or {}

    config: dict[str, Any] = defaults.get('default', {})

    if 'default' in arch_defaults:
        _dict_merge(config, arch_defaults['default'])

    if system in defaults:
        _dict_merge(config, defaults[system])

    if system in arch_defaults:
        _dict_merge(config, arch_defaults[system])

    return config


@dataclass(slots=True)
class Config:
    TRUE_VALUES: ClassVar[set[Literal['1', 'true', 'on', 'enabled', True]]] = {'1', 'true', 'on', 'enabled', True}
    MISSING: ClassVar = MISSING

    data: dict[str, Any]

    def __contains__(self, x: object, /) -> bool:
        return x in self.data

    @overload
    def get(self, key: str, default: MissingType = ..., /) -> Any | MissingType: ...

    @overload
    def get(self, key: str, default: _T, /) -> Any | _T: ...

    def get(self, key: str, default: _T | MissingType = MISSING, /) -> Any | _T | MissingType:
        return self.data.get(key, default)

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: None = None) -> bool: ...

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV]) -> _TV | _FV: ...

    def get_bool(
        self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV] | None = None
    ) -> bool | _TV | _FV:
        value = self.get(key)

        if value is MISSING:
            return default

        if isinstance(value, str):
            value = value.lower()

        if return_values is None:
            return value in self.TRUE_VALUES

        return return_values[value not in self.TRUE_VALUES]

    @overload
    def get_str(self, key: str, default: MissingType = ..., /) -> str | MissingType: ...

    @overload
    def get_str(self, key: str, default: str, /) -> str: ...

    def get_str(self, key: str, default: str | MissingType = MISSING, /) -> str | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return str(value)

    @overload
    def get_int(self, key: str, default: MissingType = ..., /) -> int | MissingType: ...

    @overload
    def get_int(self, key: str, default: int, /) -> int: ...

    def get_int(self, key: str, default: int | MissingType = MISSING, /) -> int | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return int(value)

    @overload
    def get_float(self, key: str, default: MissingType = ..., /) -> float | MissingType: ...

    @overload
    def get_float(self, key: str, default: float, /) -> float: ...

    def get_float(self, key: str, default: float | MissingType = MISSING, /) -> float | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return float(value)

    def items(self, /, *, starts_with: str | None = None) -> Iterator[tuple[str, Any]]:
        if starts_with is None:
            yield from self.data.items()
        else:
            starts_with_len = len(starts_with)
            for key, value in self.data.items():
                if key.startswith(starts_with):
                    yield key[starts_with_len:], value


def _get_rom_settings_name(rom: Path, /) -> str:
    # sanitize rule by EmulationStation
    # see FileData::getConfigurationName() on batocera-emulationstation
    rom_settings_name = rom.name.replace('=', '').replace('#', '')

    _logger.info('game settings name: %s', rom_settings_name)

    return rom_settings_name


@dataclass
class SystemConfig(Config):
    @property
    def core(self) -> str:
        return self.data['core']

    @property
    def core_forced(self) -> str:
        return self.data['core-forced']

    @property
    def emulator(self) -> str:
        return self.data['emulator']

    @property
    def emulator_forced(self) -> str:
        return self.data['emulator-forced']

    @property
    def ui_mode(self) -> Literal['Full', 'Kiosk', 'Kid']:
        return self.data['uimode']

    @property
    def show_fps(self) -> bool:
        return self.data['showFPS']

    @property
    def video_mode(self) -> str:
        return self.data.get('videomode', 'default')

    @property
    def use_guns(self) -> bool:
        return self.get_bool('use_guns')

    @property
    def use_wheels(self) -> bool:
        return self.get_bool('use_wheels')

    @classmethod
    def load(cls, options: Options, settings: UnixSettings, /) -> Self:
        defaults = _load_defaults(
            options['system'], DEFAULTS_DIR / 'configgen-defaults.yml', DEFAULTS_DIR / 'configgen-defaults-arch.yml'
        )
        data: dict[str, Any] = {
            'emulator': defaults['emulator'],
            'core': defaults['core'],
        }

        # In the yaml files, the "options" structure is not flat, so we have to flatten it here
        # because the options are flat in batocera.conf to make it easier for end users to edit
        if 'options' in defaults:
            _dict_merge(data, defaults['options'])

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        data.update(settings.get_all_iter('display', keep_name=True, keep_defaults=True))
        data.update(settings.get_all_iter('controllers', keep_name=True))

        global_settings = settings.get_all('global')
        system_settings = settings.get_all(options['system'])
        folder_settings = settings.get_all(f'{options["system"]}.folder["{rom.parent}"]')
        game_settings = settings.get_all(f'{options["system"]}["{rom_settings_name}"]')

        data.update(global_settings)
        data.update(system_settings)
        data.update(folder_settings)
        data.update(game_settings)

        try:
            es_config = ET.parse(ES_SETTINGS)

            # showFPS
            drawframerate_node = es_config.find("./bool[@name='DrawFramerate']")
            drawframerate_value = drawframerate_node.attrib['value'] if drawframerate_node is not None else 'false'
            if drawframerate_value not in ['false', 'true']:
                drawframerate_value = 'false'

            data['showFPS'] = drawframerate_value == 'true'

            # uimode
            uimode_node = es_config.find("./string[@name='UIMode']")
            uimode_value = uimode_node.attrib['value'] if uimode_node is not None else 'Full'
            if uimode_value not in ['Full', 'Kiosk', 'Kid']:
                uimode_value = 'Full'

            data['uimode'] = uimode_value
        except Exception:
            data['showFPS'] = False
            data['uimode'] = 'Full'

        data['emulator-forced'] = (
            'emulator' in global_settings
            or 'emulator' in system_settings
            or 'emulator' in game_settings
            or options['emulator'] is not None
        )
        data['core-forced'] = (
            'core' in global_settings
            or 'core' in system_settings
            or 'core' in game_settings
            or options['core'] is not None
        )

        if options['emulator'] is not None:
            data['emulator'] = options['emulator']

        if options['core'] is not None:
            data['core'] = options['core']

        if 'use_guns' not in data and options['lightgun']:
            data['use_guns'] = True

        if 'use_wheels' not in data and options['wheel']:
            data['use_wheels'] = True

        # network options
        if options['netplaymode'] is not None:
            data['netplay.mode'] = options['netplaymode']

        if options['netplaypass'] is not None:
            data['netplay.password'] = options['netplaypass']

        if options['netplayip'] is not None:
            data['netplay.server.ip'] = options['netplayip']

        if options['netplayport'] is not None:
            data['netplay.server.port'] = options['netplayport']

        if options['netplaysession'] is not None:
            data['netplay.server.session'] = options['netplaysession']

        # autosave arguments
        if options['state_slot'] is not None:
            data['state_slot'] = options['state_slot']

        if options['autosave'] is not None:
            data['autosave'] = options['autosave']

        if options['state_filename'] is not None:
            data['state_filename'] = options['state_filename']

        return cls(data)


@dataclass
class RenderConfig(Config):
    @classmethod
    def load(cls, options: Options, settings: UnixSettings, shader_set: str | MissingType, /) -> Self:
        data: dict[str, Any] = {}

        if shader_set is not cls.MISSING:
            if shader_set == 'none':
                rendering_defaults = BATOCERA_SHADERS / 'configs' / 'rendering-defaults.yml'
            else:
                rendering_defaults = USER_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'
                if not rendering_defaults.exists():
                    rendering_defaults = BATOCERA_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'

            data = _load_defaults(
                options['system'], rendering_defaults, rendering_defaults.with_name('rendering-defaults-arch.yml')
            )

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        # es only allow to update system and game settings in fact for the moment

        # for compatibility with earlier Batocera versions, let's keep -renderer
        # but it should be reviewed when we refactor configgen (to Python3?)
        # so that we can fetch them from system.shader without -renderer
        data.update(settings.get_all_iter(f'{options["system"]}-renderer'))
        data.update(settings.get_all_iter(f'{options["system"]}["{rom_settings_name}"]-renderer'))

        return cls(data)
