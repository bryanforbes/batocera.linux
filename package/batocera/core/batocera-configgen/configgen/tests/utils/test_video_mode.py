from __future__ import annotations

import subprocess
from typing import TYPE_CHECKING, Any
from unittest.mock import call

import pytest

from configgen.config import SystemConfig
from configgen.exceptions import BatoceraException
from configgen.utils.videoMode import (
    changeMode,
    changeMouse,
    checkModeExists,
    getAltDecoration,
    getCurrentMode,
    getCurrentOutput,
    getCurrentResolution,
    getGLVendor,
    getGLVersion,
    getRefreshRate,
    getScreens,
    getScreensInfos,
    isResolutionReversed,
    minTomaxResolution,
    supportSystemRotation,
)

if TYPE_CHECKING:
    from unittest.mock import Mock

    from pyfakefs.fake_filesystem import FakeFilesystem
    from pytest_mock import MockerFixture
    from syrupy.assertion import SnapshotAssertion

    from configgen.types import Resolution, ScreenInfo


@pytest.fixture
def check_mode_exists(mocker: MockerFixture) -> Mock:
    return mocker.patch('configgen.utils.videoMode.checkModeExists')


def test_change_mode(subprocess_run: Mock, check_mode_exists: Mock) -> None:
    check_mode_exists.return_value = True

    changeMode('1920x1080.120.0')

    subprocess_run.assert_called_once_with(
        ['batocera-resolution', 'setMode', '1920x1080.120.0'], capture_output=True, text=True, check=True
    )


def test_change_mode_check_fails(subprocess_run: Mock, check_mode_exists: Mock) -> None:
    check_mode_exists.return_value = False

    changeMode('1920x1080.120.0')

    subprocess_run.assert_not_called()


def test_change_mode_first_fails(mocker: MockerFixture, subprocess_run: Mock, check_mode_exists: Mock) -> None:
    check_mode_exists.return_value = True

    subprocess_run.side_effect = [
        subprocess.CalledProcessError(1, ['batocera-resolution', 'setMode', '1920x1080.120.0']),
        mocker.Mock(),
    ]

    changeMode('1920x1080.120.0')

    assert subprocess_run.call_args_list == [
        call(['batocera-resolution', 'setMode', '1920x1080.120.0'], capture_output=True, text=True, check=True),
        call(['batocera-resolution', 'setMode', '1920x1080.120.0'], capture_output=True, text=True, check=True),
    ]


def test_change_mode_fails(subprocess_run: Mock, check_mode_exists: Mock) -> None:
    check_mode_exists.return_value = True

    subprocess_run.side_effect = subprocess.CalledProcessError(1, ['batocera-resolution', 'setMode', '1920x1080.120.0'])

    with pytest.raises(BatoceraException, match=r'^Error setting video mode$'):
        changeMode('1920x1080.120.0')


def test_get_current_mode(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (
        b"""1920x1080.120.00
1024x768.120.00
""",
        b'',
    )
    subprocess_popen.return_value = popen_instance

    assert getCurrentMode() == '1920x1080.120.00'
    subprocess_popen.assert_called_once_with(['batocera-resolution currentMode'], stdout=subprocess.PIPE, shell=True)


def test_get_current_mode_none(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (b'', b'')
    subprocess_popen.return_value = popen_instance

    assert getCurrentMode() is None
    subprocess_popen.assert_called_once_with(['batocera-resolution currentMode'], stdout=subprocess.PIPE, shell=True)


def test_get_refresh_rate(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (
        b"""120.00
60.00
""",
        b'',
    )
    subprocess_popen.return_value = popen_instance

    assert getRefreshRate() == '120.00'
    subprocess_popen.assert_called_once_with(['batocera-resolution refreshRate'], stdout=subprocess.PIPE, shell=True)


def test_get_refresh_rate_none(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (b'', b'')
    subprocess_popen.return_value = popen_instance

    assert getRefreshRate() is None
    subprocess_popen.assert_called_once_with(['batocera-resolution refreshRate'], stdout=subprocess.PIPE, shell=True)


@pytest.mark.parametrize(
    ('screens', 'output', 'resolutions', 'config', 'expected', 'expected_calls'),
    [
        pytest.param(
            ['screen1'],
            'screen1',
            [{'width': 1920, 'height': 1080}],
            {},
            [{'width': 1920, 'height': 1080, 'x': 0, 'y': 0}],
            [call()],
            id='one screen',
        ),
        pytest.param(
            ['screen1', 'screen2'],
            'screen2',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}],
            {},
            [{'width': 640, 'height': 480, 'x': 0, 'y': 0}, {'width': 1920, 'height': 1080, 'x': 640, 'y': 0}],
            [call(), call('screen1')],
            id='two screens',
        ),
        pytest.param(
            ['screen1', 'screen2'],
            'screen2',
            [{'width': 640, 'height': 480}, Exception],
            {},
            [{'width': 640, 'height': 480, 'x': 0, 'y': 0}],
            [call(), call('screen1')],
            id='two screens, screen1 resolution raises',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3'],
            'screen3',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, {'width': 1024, 'height': 768}],
            {},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
                {'width': 1024, 'height': 768, 'x': 2560, 'y': 0},
            ],
            [call(), call('screen1'), call('screen2')],
            id='three screens',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3'],
            'screen3',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, Exception],
            {},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
            ],
            [call(), call('screen1'), call('screen2')],
            id='three screens, screen2 resolution raises',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3'],
            'screen2',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, {'width': 1024, 'height': 768}],
            {'videooutput2': 'screen3'},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
                {'width': 1024, 'height': 768, 'x': 2560, 'y': 0},
            ],
            [call(), call('screen3'), call('screen1')],
            id='three screens, videooutput2 set to screen3',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3'],
            'screen2',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, {'width': 1024, 'height': 768}],
            {'videooutput2': 'screen2'},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
                {'width': 1024, 'height': 768, 'x': 2560, 'y': 0},
            ],
            [call(), call('screen1'), call('screen3')],
            id='three screens, videooutput2 set to output 1',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3'],
            'screen2',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, {'width': 1024, 'height': 768}],
            {'videooutput2': 'screen4'},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
                {'width': 1024, 'height': 768, 'x': 2560, 'y': 0},
            ],
            [call(), call('screen1'), call('screen3')],
            id='three screens, videooutput2 set to screen4',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3', 'screen4'],
            'screen2',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, {'width': 1024, 'height': 768}],
            {'videooutput3': 'screen4'},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
                {'width': 1024, 'height': 768, 'x': 2560, 'y': 0},
            ],
            [call(), call('screen1'), call('screen4')],
            id='four screens, videooutput3 set to screen4',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3', 'screen4'],
            'screen2',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, {'width': 1024, 'height': 768}],
            {'videooutput3': 'screen1'},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
                {'width': 1024, 'height': 768, 'x': 2560, 'y': 0},
            ],
            [call(), call('screen1'), call('screen3')],
            id='four screens, videooutput3 set to screen1',
        ),
        pytest.param(
            ['screen1', 'screen2', 'screen3', 'screen4'],
            'screen2',
            [{'width': 640, 'height': 480}, {'width': 1920, 'height': 1080}, {'width': 1024, 'height': 768}],
            {'videooutput2': 'screen3', 'videooutput3': 'screen1'},
            [
                {'width': 640, 'height': 480, 'x': 0, 'y': 0},
                {'width': 1920, 'height': 1080, 'x': 640, 'y': 0},
                {'width': 1024, 'height': 768, 'x': 2560, 'y': 0},
            ],
            [call(), call('screen3'), call('screen1')],
            id='four screens, videooutput2 set to screen3, videooutput3 set to screen1',
        ),
    ],
)
def test_get_screens_infos(
    mocker: MockerFixture,
    screens: list[str],
    output: str,
    resolutions: list[Resolution | type[Exception]],
    config: dict[str, str],
    expected: list[ScreenInfo],
    expected_calls: list[Any],
) -> None:
    mocker.patch('configgen.utils.videoMode.getScreens', return_value=screens)
    mocker.patch('configgen.utils.videoMode.getCurrentOutput', return_value=output)
    get_current_resolution = mocker.patch('configgen.utils.videoMode.getCurrentResolution', side_effect=resolutions)

    assert getScreensInfos(SystemConfig(config)) == expected
    assert get_current_resolution.call_args_list == expected_calls


def test_get_screens(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (
        b"""foo
bar
""",
        b'',
    )
    subprocess_popen.return_value = popen_instance

    assert getScreens() == ['foo', 'bar']
    subprocess_popen.assert_called_once_with(['batocera-resolution listOutputs'], stdout=subprocess.PIPE, shell=True)


def test_min_to_max_resolution(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (b'', b'')
    subprocess_popen.return_value = popen_instance

    minTomaxResolution()

    subprocess_popen.assert_called_once_with(
        ['batocera-resolution minTomaxResolution'], stdout=subprocess.PIPE, shell=True
    )
    popen_instance.communicate.assert_called_once_with()


def test_get_current_resolution(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (b'1920x1080\n', b'')
    subprocess_popen.return_value = popen_instance

    assert getCurrentResolution() == {'width': 1920, 'height': 1080}
    subprocess_popen.assert_called_once_with(
        ['batocera-resolution currentResolution'], stdout=subprocess.PIPE, shell=True
    )


def test_get_current_resolution_for_screen(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (b'640x480\n', b'')
    subprocess_popen.return_value = popen_instance

    assert getCurrentResolution('screen_one') == {'width': 640, 'height': 480}
    subprocess_popen.assert_called_once_with(
        ['batocera-resolution --screen screen_one currentResolution'], stdout=subprocess.PIPE, shell=True
    )


def test_get_current_output(mocker: MockerFixture, subprocess_popen: Mock) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (
        b' \t \n screen_one \t \r   \n',
        b'',
    )
    subprocess_popen.return_value = popen_instance

    assert getCurrentOutput() == 'screen_one'
    subprocess_popen.assert_called_once_with(['batocera-resolution currentOutput'], stdout=subprocess.PIPE, shell=True)


@pytest.mark.parametrize(('returncode', 'expected'), [(0, True), (1, False)])
def test_support_system_rotation(
    mocker: MockerFixture, subprocess_popen: Mock, returncode: int, expected: bool
) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (b'', b'')
    popen_instance.returncode = returncode
    subprocess_popen.return_value = popen_instance

    assert supportSystemRotation() == expected
    subprocess_popen.assert_called_once_with(
        ['batocera-resolution supportSystemRotation'], stdout=subprocess.PIPE, shell=True
    )
    popen_instance.communicate.assert_called_once_with()


@pytest.mark.parametrize('expected', [True, False])
def test_is_resolution_reversed(fs: FakeFilesystem, expected: bool) -> None:
    if expected:
        fs.create_file('/var/run/rk-rotation')

    assert isResolutionReversed() == expected


@pytest.mark.parametrize(
    ('mode', 'expected'),
    [
        ('max-1920x1080', True),
        ('1920x1080.120.00', True),
        ('640x480.60.00', False),
        ('max-foo', False),
    ],
)
def test_check_mode_exists(
    mocker: MockerFixture, subprocess_popen: Mock, mode: str, expected: bool, snapshot: SnapshotAssertion
) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (
        b"""max-1920x1080:maximum 1920x1080
1920x1080.120.00:1920x1080 120.00 Hz
1024x768.120.00:1024x768 120.00 Hz
640x480.120.00:640x480 120.00 Hz
""",
        b'',
    )
    subprocess_popen.return_value = popen_instance

    assert checkModeExists(mode) == expected
    assert subprocess_popen.call_args_list == snapshot


@pytest.mark.parametrize(('mode', 'expected'), [(True, 'show'), (False, 'hide')])
def test_change_mouse(mocker: MockerFixture, subprocess_popen: Mock, mode: bool, expected: str) -> None:
    popen_instance = mocker.Mock()
    popen_instance.communicate.return_value = (b'', b'')
    subprocess_popen.return_value = popen_instance

    changeMouse(mode)
    subprocess_popen.assert_called_once_with([f'batocera-mouse {expected}'], stdout=subprocess.PIPE, shell=True)
    popen_instance.communicate.assert_called_once_with()


@pytest.mark.parametrize(('version', 'expected'), [('4.6', 4.6), ('3.2.3.0', 3.2)])
def test_get_gl_version(fs: FakeFilesystem, mocker: MockerFixture, version: str, expected: float) -> None:
    fs.create_file('/usr/bin/glxinfo')

    check_output = mocker.patch(
        'subprocess.check_output',
        return_value=b'OpenGL version string: ' + version.encode('utf-8') + b' (Compatibility Profile) Mesa 24.3.3',
    )

    assert getGLVersion() == expected
    check_output.assert_called_once_with('glxinfo | grep "OpenGL version"', shell=True)


def test_get_gl_version_no_glxinfo(mocker: MockerFixture) -> None:
    check_output = mocker.patch('subprocess.check_output', return_value=b'')

    assert getGLVersion() == 0
    check_output.assert_not_called()


def test_get_gl_version_exception(fs: FakeFilesystem, mocker: MockerFixture) -> None:
    fs.create_file('/usr/bin/glxinfo')
    mocker.patch('subprocess.check_output', side_effect=Exception)

    assert getGLVersion() == 0


def test_get_gl_vendor(fs: FakeFilesystem, mocker: MockerFixture) -> None:
    fs.create_file('/usr/bin/glxinfo')

    check_output = mocker.patch(
        'subprocess.check_output',
        return_value=b'OpenGL vendor string: AmD',
    )

    assert getGLVendor() == 'amd'
    check_output.assert_called_once_with('glxinfo | grep "OpenGL vendor string"', shell=True)


def test_get_gl_vendor_no_glxinfo(mocker: MockerFixture) -> None:
    check_output = mocker.patch('subprocess.check_output', return_value=b'')

    assert getGLVendor() == 'unknown'
    check_output.assert_not_called()


def test_get_gl_vendor_exception(fs: FakeFilesystem, mocker: MockerFixture) -> None:
    fs.create_file('/usr/bin/glxinfo')
    mocker.patch('subprocess.check_output', side_effect=Exception)

    assert getGLVendor() == 'unknown'


@pytest.mark.parametrize('emulator', ['flycast', 'mame', 'retroarch'])
@pytest.mark.parametrize('system', ['dreamcast', 'naomi'])
def test_get_alt_decoration(system: str, emulator: str) -> None:
    assert getAltDecoration(system, '/path/to/rom.zip', emulator) == ('standalone' if emulator == 'flycast' else '0')


@pytest.mark.parametrize('emulator', ['mame', 'retroarch'])
@pytest.mark.parametrize(
    'system', ['lynx', 'wswan', 'wswanc', 'mame', 'fbneo', 'naomi', 'atomiswave', 'nds', '3ds', 'vectrex']
)
def test_get_alt_decoration_special_file(fs: FakeFilesystem, system: str, emulator: str) -> None:
    fs.create_file(
        f'/usr/share/batocera/configgen/data/special/{system}.csv',
        contents="""foo - bar;270
rOm;90
baz blah ham (Spam);180
""",
    )

    assert getAltDecoration(system, '/path/to/RoM.zip', emulator) == '90'


@pytest.mark.parametrize('emulator', ['mame', 'retroarch'])
@pytest.mark.parametrize(
    'system', ['lynx', 'wswan', 'wswanc', 'mame', 'fbneo', 'naomi', 'atomiswave', 'nds', '3ds', 'vectrex']
)
def test_get_alt_decoration_special_file_not_found(fs: FakeFilesystem, system: str, emulator: str) -> None:
    fs.create_file(
        f'/usr/share/batocera/configgen/data/special/{system}.csv',
        contents="""foo - bar;270
baz blah ham (Spam);180
""",
    )

    assert getAltDecoration(system, '/path/to/RoM.zip', emulator) == '0'
