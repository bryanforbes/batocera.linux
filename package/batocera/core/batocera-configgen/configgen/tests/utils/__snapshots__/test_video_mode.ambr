# serializer version: 1
# name: test_check_mode_exists[1920x1080.120.00-True]
  _CallList([
    _Call(
      tuple(
        list([
          'batocera-resolution listModes',
        ]),
      ),
      dict({
        'shell': True,
        'stdout': -1,
      }),
    ),
  ])
# ---
# name: test_check_mode_exists[640x480.60.00-False]
  _CallList([
    _Call(
      tuple(
        list([
          'batocera-resolution listModes',
        ]),
      ),
      dict({
        'shell': True,
        'stdout': -1,
      }),
    ),
  ])
# ---
# name: test_check_mode_exists[max-1920x1080-True]
  _CallList([
  ])
# ---
# name: test_check_mode_exists[max-foo-False]
  _CallList([
    _Call(
      tuple(
        list([
          'batocera-resolution listModes',
        ]),
      ),
      dict({
        'shell': True,
        'stdout': -1,
      }),
    ),
  ])
# ---
